### PR TITLE
Nested neighbour reductions

### PR DESCRIPTION
**Pull request by cosunae**
Description
===========

The original implementation for parsing and code gen the neighbur reductions relied on a map:
`std::map<int, std::stringstream> reductionMap_`
https://github.com/MeteoSwiss-APN/dawn/pull/1135/files#diff-0c1971e75cf42c11b0563ba98a06373a2888735a5f5953e8024c90917f1f4e6bL91
that stored for each expression id of a neighbour reduction the stream with codegen of the neighbour loop structure. 

It required two passes of the ASTStencilBody, one to fill all the streams within this map and a second in order to insert them into the main stream of the ASTStencilBody. 

This design was assuming there are no nested neighbour reductions.  
In the process of generating the stream, first we create the neighbour for loop and indices and then it generates the expression (RHS) of the ast::ReductionOverNeighborExpr. If another ast::ReductionOverNeighborExpr is encountered, we should generate another block for loop, but now it is too late, since we are already generating the statement. 
Following the original strategy we should parse the ASTStencilBody not two times, but N times instead, where N is the number of nested reductions +1.

Here instead we propose a different design, recursive to fit better with the nature of nested reductions. 
Each reduction has an associated ASTStencilBody: `std::map<int, std::unique_ptr<ASTStencilBody>> reductionParser_`
https://github.com/MeteoSwiss-APN/dawn/pull/1135/files#diff-0c1971e75cf42c11b0563ba98a06373a2888735a5f5953e8024c90917f1f4e6bR73

That ASTStencilBody will exclusively parse that neighbour reduction. Recursively it parses its children neighbour red of the RHS into reductionParser recursively creating a tree of ASTStencilBody. 
Traversing that tree we resolve and insert the code gen of each neighbour reduction level into the main ss_ of the root ASTStencilBody

Tested with
========== 
https://github.com/dawn-ico/icondusk-e2e/pull/45

### CODE REVIEW

File: dawn/src/dawn/CodeGen/Cuda-ico/ASTStencilBody.cpp [hunk 1]
```diff
@@ -273,111 +285,95 @@ bool ASTStencilBody::hasIrregularPentagons(const std::vector<ast::LocationType>&
...
+    return nbhIdx_m1Str();
+  else
+    return "pidx";
+}
 
-  if(!firstPass_) {
-    ss_ << lhs_name;
-    return;
-  }
+void ASTStencilBody::evalNeighbourReductionLambda(
```
Comment (mroethlin) [inline dawn/src/dawn/CodeGen/Cuda-ico/ASTStencilBody.cpp:300]: Remainder of the lambda reduction strategy?

**Review by mroethlin — CHANGES_REQUESTED** [review]
I think there is a problem with the generation of the (nested) indices. 

```
@stencil
def nested(a: Field[Cell], b: Field[Edge], c: Field[Vertex], d: Field[Edge]):
    with domain.upward:
        a = sum_over(Cell > Edge , b * sum_over(Edge > Vertex, c*sum_over(Vertex > Edge, d)))
```

is translated to

```
  for(int kIter = klo; kIter < khi; kIter++) {
    if(kIter >= kSize + 0) {
      return;
    }
    ::dawn::float_type lhs_69 = (::dawn::float_type)0;
    for(int nbhIter0 = 0; nbhIter0 < C_E_SIZE; nbhIter0++) {
      int nbhIdx0 = ceTable[pidx * C_E_SIZE + nbhIter0];
      if(nbhIdx0 == DEVICE_MISSING_VALUE) {
        continue;
      }
      ::dawn::float_type lhs_66 = (::dawn::float_type)0;
      for(int nbhIter0 = 0; nbhIter0 < E_V_SIZE; nbhIter0++) {
        int nbhIdx0 = evTable[pidx * E_V_SIZE + nbhIter0];
        if(nbhIdx0 == DEVICE_MISSING_VALUE) {
          continue;
        }
        ::dawn::float_type lhs_63 = (::dawn::float_type)0;
        for(int nbhIter0 = 0; nbhIter0 < V_E_SIZE; nbhIter0++) {
          int nbhIdx0 = veTable[pidx * V_E_SIZE + nbhIter0];
          if(nbhIdx0 == DEVICE_MISSING_VALUE) {
            continue;
          }
          lhs_63 += d[nbhIdx0];
        }
        lhs_66 += (c[nbhIdx0] * lhs_63);
      }
      lhs_69 += (b[nbhIdx0] * lhs_66);
    }
    a[pidx] = lhs_69;
  }
```

note that all index names have the `0` suffix. I think this may be correct for that case because of the c++ scoping rules. However, in the function `pidx()` you attempt to retrieve the parent index (which I can't seem to trigger btw), indicating that naming all indices the same wasn't your intention.

File: dawn/src/dawn/CodeGen/Cuda-ico/ASTStencilBody.h [hunk 2]
```diff
@@ -82,37 +68,37 @@ class ASTStencilBody : public ASTCodeGenCXX {
...
-  ASTStencilBody(const iir::StencilMetaInformation& metadata, bool genAtlasCompatCode);
 
-  virtual ~ASTStencilBody();
+  ASTStencilBody(const iir::StencilMetaInformation& metadata, bool genAtlasCompatCode,
+                 int recursiveIterNest = 0);
 
-  void setFirstPass() { firstPass_ = true; };
-  void setSecondPass() { firstPass_ = false; };
-  void setBlockID(int currentBlock) { currentBlock_ = currentBlock; }
-  void setBlockToMergeGroupMap(std::optional<MergeGroupMap> blockToMergeGroupMap) {
```
[the remark this answers is not in the document]
Comment (cosunae) [reply dawn/src/dawn/CodeGen/Cuda-ico/ASTStencilBody.h:113]: indeed, I missed that one. I removed it

**Comment by cosunae** [comment]
@mroethlin this
https://github.com/MeteoSwiss-APN/dawn/pull/1135/commits/b713cb07043dd85294bee5ce5c0a3c7fad98a16b
fixes the issue that all the nested for loops are generated with the same index for iterators. It was a problem with a merge, that changed the ctr

File: dawn/src/dawn/CodeGen/Cuda-ico/ASTStencilBody.h [hunk 3]
```diff
@@ -82,37 +68,37 @@ class ASTStencilBody : public ASTCodeGenCXX {
...
   bool hasIrregularPentagons(const std::vector<ast::LocationType>& chain);
+  void evalNeighbourReduction(const std::shared_ptr<ast::ReductionOverNeighborExpr>& expr);
+  void generateNeighbourRedLoop(std::stringstream& ss) const;
+  std::string nbhLhsName(const std::shared_ptr<ast::Expr>& expr);
+  std::string pidxStr();
+  std::string nbhIterStr();
+  // symbol of the current neighbour index in a reduction loop
+  std::string nbhIdxStr();
+  // in case of nested reduction, this is the symbol of the index of parent reduction loop
+  std::string nbhIdxParentStr();
```
Comment (mroethlin) [inline dawn/src/dawn/CodeGen/Cuda-ico/ASTStencilBody.h:90]: I think those getters should be marked `const`

File: dawn/src/dawn/CodeGen/Cuda-ico/ASTStencilBody.h [hunk 4]
```diff
@@ -39,36 +39,22 @@ class StencilMetaInformation;
...
 public:
   void visit(const std::shared_ptr<ast::ReductionOverNeighborExpr>& expr) override {
-    foundReductions_.push_back(expr);
-    reductionToBlock_.insert({expr->getID(), currentBlock_});
+    found_ = true;
     return;
   }
   void visit(const std::shared_ptr<ast::BlockStmt>& stmt) override {
-    currentBlock_ = stmt->getID();
+    // TODO do we need this override ?
```
Comment (mroethlin) [inline dawn/src/dawn/CodeGen/Cuda-ico/ASTStencilBody.h:51]: This looks like an exact code clone of `ASTVISITORFORWARDING_VISIT_IMPL` to me, should be able to safely remove this

**Review by mroethlin — APPROVED** [review]
Two very minor comments. Otherwise gtg after CI.

File: dawn/src/dawn/CodeGen/Cuda-ico/ASTStencilBody.h (hunk 3 above)
Comment (cosunae) [reply dawn/src/dawn/CodeGen/Cuda-ico/ASTStencilBody.h:90]: done

File: dawn/src/dawn/CodeGen/Cuda-ico/ASTStencilBody.h (hunk 4 above)
Comment (cosunae) [reply dawn/src/dawn/CodeGen/Cuda-ico/ASTStencilBody.h]: done

**Comment by cosunae** [comment]
test passing here: 
https://gitlab.dkrz.de/dsl/icon-cscs/-/merge_requests/29
and here
https://github.com/dawn-ico/icondusk-e2e/pull/45

**Pull request merged**